### PR TITLE
Added comments for pulsesensor and adc

### PR DESCRIPTION
**Pull request by Fred194**

### CODE REVIEW

File: adc/adc.c [hunk 1]
```diff
@@ -6,31 +6,34 @@
 #include <avr/io.h>
 #include <avr/interrupt.h>
 
-int adc_current_value = 0;
+int adc_current_value = 0; //Integer to store the currently read adc value
```
Comment (Max-Leopold) [inline adc/adc.c:9]: Bitte noch ein bisschen genauer, was genau dieser Wert ist und wofür er gebraucht wird

File: adc/adc.c [hunk 2]
```diff
@@ -6,31 +6,34 @@
...
     ADMUX &= ~(1 << ADLAR);   // clear for 10 bit resolution
 
-    ADCSRA |= (1 << ADPS2) | (1 << ADPS1) | (1 << ADPS0);    // 128 prescale for 16Mhz
+    ADCSRA |= (1 << ADPS2) | (1 << ADPS1) | (1 << ADPS0);    // Use 128 prescaler for 16Mhz
     ADCSRA |= (1 << ADATE);   // Set ADC Auto Trigger Enable
-    ADCSRB = 0;               // 0 for free running mode
-    ADCSRA |= (1 << ADEN);    // Enable the ADC
+    ADCSRB = 0;               // 0 for free running mode - free running means: The ADC starts automatically a new measurement after the last one
+    ADCSRA |= (1 << ADEN);    // Must be set to enable the ADC
     ADCSRA |= (1 << ADIE);    // Enable Interrupts
```
Comment (Max-Leopold) [inline adc/adc.c:22]: Enable ADC Interrupts. Maybe ein bisschen präziser

File: adc/adc.c [hunk 3]
```diff
@@ -6,31 +6,34 @@
...
 int get_adc_current(){
 	return adc_current_value;
 }
 
+//Interrupt for the ADC
 ISR (ADC_vect)
 {
-	int ADCval = ADCL;
-	ADCval = (ADCH << 8) + ADCval;
+    //When a conversion is complete, the measured value is in these two registers (ADCL and ADCH)
```
Comment (Max-Leopold) [inline adc/adc.c:34]: the converted value. Not the measured one

File: pulsesensor/pulsesensor.c [hunk 4]
```diff
@@ -13,21 +13,21 @@
...
-int sum = 0, bpm = 0;
-int arrayLength = 0;
-int waitTime = 60;
-unsigned long startMillis;
-unsigned long sendStartMillis;
-long lastHearbeatdetected;
-
-
-
+int valueLength = 8;        //Integer to set the length in the loops, so that you don't always have to change the number in the loops manually
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:16]: Does not really explain what this variable is for

File: pulsesensor/pulsesensor.c [hunk 5]
```diff
@@ -13,21 +13,21 @@
...
+int bpmValues[8];           //To set the array length
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:17]: nein, dass ist nicht die array length. Das ist ein array

File: pulsesensor/pulsesensor.c [hunk 6]
```diff
@@ -13,21 +13,21 @@
...
+int average = 0;            //Integer to store the average of the read bpm values
+int sum = 0, bpm = 0;       //Integer to store the sum of the bpm´s and Integer to store the bpm value
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:19]: please split this in two declarations

File: pulsesensor/pulsesensor.c [hunk 7]
```diff
@@ -13,21 +13,21 @@
...
+int arrayLength = 0;        //Integer that stores the length of the bpmValues-Array (to calculate the average correctly)
+int waitTime = 60;          //Integer to store the wait time for debounce purposes
+unsigned long startMillis;  //Timestamps variable for delay purposes
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:22]: bischen genauer alles erklären bitte. Du darfst auch überall mehrzeilige kommentare einfügen

File: pulsesensor/pulsesensor.c [hunk 8]
```diff
@@ -37,36 +37,36 @@ void pulsesensor_main() {
...
 
     /* if-statement in which we get the information if a heartbeat is detected or not
-    ---> if yes: the value gets added to the bpmValues-array via the addValue-method
+    ---> if yes: the value gets added to the bpmValues-Array via the addValue-method
     ---> if the sensor doesn't recognize a signal within 7 seconds the output will be 0 */
 
-    int elapsedMillis = getMillis() - startMillis;
-    if (elapsedMillis >= waitTime){
+    int elapsedMillis = getMillis() - startMillis;      //To calculate the elapsed time
+    if (elapsedMillis >= waitTime){                     //For debounce purposes
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:44]: genauer bitte

File: pulsesensor/pulsesensor.c [hunk 9]
```diff
@@ -98,40 +98,41 @@ char heartbeatDetected(int delay, int ADCvalue) {
...
-
+//Returns the average of the bpm values
 int getBpm() {
     return average;
 }
-
+//Adds bpm value to the bpmValues-Array
 void addValue(int bpm) {
-    //for normalizing purpose
+    //For normalizing purpose - values under 40 and over 200 will not be saved in the bpmValues-Array
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:107]: schreib lieber sowas wie: "Normalize the values ...". Nicht immer "purpose"

File: adc/adc.c (hunk 1 above)
Comment (Fred194) [reply adc/adc.c]: Habs genauer beschrieben

File: adc/adc.c (hunk 2 above)
Comment (Fred194) [reply adc/adc.c]: Habs präziser beschrieben

File: adc/adc.c (hunk 3 above)
Comment (Fred194) [reply adc/adc.c:35]: Changed this

File: pulsesensor/pulsesensor.c (hunk 4 above)
Comment (Fred194) [reply pulsesensor/pulsesensor.c]: Added detailed description

File: pulsesensor/pulsesensor.c (hunk 5 above)
Comment (Fred194) [reply pulsesensor/pulsesensor.c]: Habe dies korrigiert

File: pulsesensor/pulsesensor.c (hunk 6 above)
Comment (Fred194) [reply pulsesensor/pulsesensor.c]: Splitted these declarations

File: pulsesensor/pulsesensor.c (hunk 7 above)
Comment (Fred194) [reply pulsesensor/pulsesensor.c]: Habe dies nun genauer erklärt

File: pulsesensor/pulsesensor.c (hunk 8 above)
Comment (Fred194) [reply pulsesensor/pulsesensor.c]: Habe dies nun genauer erklärt

File: pulsesensor/pulsesensor.c (hunk 9 above)
Comment (Fred194) [reply pulsesensor/pulsesensor.c]: Habe den Vorschlag eingebaut

**Comment by Max-Leopold** [comment]
Ich glaube du hast irgendwas nicht richtig gemacht oder ich bin blind. Aber du hast doch gar nichts geändert oder?

**Comment by Max-Leopold** [comment]
Nope, immer noch nicht. Ich glaub du hast den merge falsch gemacht

**Comment by Fred194** [comment]
somehow it broke the commitment. had to search for the changes myself. They are all in the "Added comments" https://github.com/Max-Leopold/MeBike/pull/36/commits/40d419567f802bd3c5c430fa75bd3850da6f016f

**Comment by Max-Leopold** [comment]
If I can't find them under "Filex changed", they wont be included if you merge this to master

**Comment by Fred194** [comment]
The latest version of the code with comments are now visible at "Files changed"

File: pulsesensor/pulsesensor.c [hunk 10]
```diff
@@ -13,21 +13,26 @@
...
-int sum = 0, bpm = 0;
-int arrayLength = 0;
-int waitTime = 60;
-unsigned long startMillis;
-unsigned long sendStartMillis;
-long lastHearbeatdetected;
-
-
 
+int valueLength = 8;        /* Defines the length of the bpmValues-Array instead of using "array.length"-method.
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:17]: Check this post https://stackoverflow.com/questions/37538/how-do-i-determine-the-size-of-my-array-in-c

File: adc/adc.c [hunk 11]
```diff
@@ -6,31 +6,35 @@
...
 
-    ADCSRA |= (1 << ADPS2) | (1 << ADPS1) | (1 << ADPS0);    // 128 prescale for 16Mhz
+    ADCSRA |= (1 << ADPS2) | (1 << ADPS1) | (1 << ADPS0);    // Use 128 prescaler for 16Mhz
     ADCSRA |= (1 << ADATE);   // Set ADC Auto Trigger Enable
-    ADCSRB = 0;               // 0 for free running mode
-    ADCSRA |= (1 << ADEN);    // Enable the ADC
-    ADCSRA |= (1 << ADIE);    // Enable Interrupts
+    ADCSRB = 0;               // 0 for free running mode - free running means: The ADC starts automatically a new measurement after the last one
+    ADCSRA |= (1 << ADEN);    // Must be set to enable the ADC
+    ADCSRA |= (1 << ADIE);    // Enable Interrupts: If a 1 is written to ADIE, the ADC triggers an interrupt after completion of a measurement.
```
Comment (Max-Leopold) [inline adc/adc.c:23]: I dont really understand this comment.

File: adc/adc.c [hunk 12]
```diff
@@ -6,31 +6,35 @@
...
 }
 
+//Interrupt for the ADC
 ISR (ADC_vect)
 {
-	int ADCval = ADCL;
-	ADCval = (ADCH << 8) + ADCval;
+    //When a conversion is complete, the converted value is in these two registers (ADCL and ADCH)
+	int ADCval = ADCL; //Integer to store the value from register ADCL
+	ADCval = (ADCH << 8) + ADCval; //To read the value of the ADCH register plus the value of ADCL
```
Comment (Max-Leopold) [inline adc/adc.c:37]: What does this lead to?

Comment (Max-Leopold) [reply adc/adc.c:37]: Why are we doing this?

File: pulsesensor/pulsesensor.c [hunk 13]
```diff
@@ -13,21 +13,26 @@
...
+                             Is used in the loops, so it´s not hardcoded and can be changed easily. */
+int bpmValues[8];           //Array with a length of 8 to store the bpm values in it
+int average = 0;            /* Integer to store the average of the read bpm values
+                             * Is used to display the average of the last 8 read bpm values */
+int sum = 0;                /* Integer to store the sum of the bpm´s to calculate the average
+                             * The last 8 bpm values ​​are added up here                                             */
+int bpm = 0;                //Integer to store value of the beats per minute.
+int arrayLength = 0;        //Integer the stores the amount of the bpm values in the array to calculate the average
+int waitTime = 60;          //Integer to store the wait time to realize delay functionality
+unsigned long startMillis;  //Timestamps variable to realize delay functionality. For example is used to calculate the elapsed time
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:27]: For example it is used...

File: pulsesensor/pulsesensor.c [hunk 14]
```diff
@@ -13,21 +13,26 @@
...
+unsigned long sendStartMillis; //Timestamps variable for delay purposes
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:28]: delay purposes for what?

File: pulsesensor/pulsesensor.c [hunk 15]
```diff
@@ -13,21 +13,26 @@
...
+long lastHearbeatdetected;  //To store the time the last heartbeat was detected
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:29]: Why you need to store this value?

File: pulsesensor/pulsesensor.c [hunk 16]
```diff
@@ -13,21 +13,26 @@
...
+
+
+//Init method to initialize
 void pulsesensor_init() {
-    startMillis = getMillis();
-	sendStartMillis = getMillis();
+    startMillis = getMillis();      //Set the initial time for the pulse sensor
+	sendStartMillis = getMillis();  //Set the inital time for the bluetooth sensor
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:35]: Why do you need this time twice? Wouldn't one variable be enough?

File: pulsesensor/pulsesensor.c [hunk 17]
```diff
@@ -37,36 +42,39 @@ void pulsesensor_main() {
...
-    if (elapsedMillis >= waitTime){
+
+    int elapsedMillis = getMillis() - startMillis;      /*To calculate the elapsed time between the current time
+                                                         * (getMillis) and the the measured time (startMillis) */
+    if (elapsedMillis >= waitTime){                     /* Query if the elapsed time is greater than or equal to the
+                                                         * waiting time to get delay functionality */
         startMillis = getMillis();
         static int beatMsec = 0;
 
         if (heartbeatDetected(waitTime, get_adc_current())) {
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:56]: What is done here?

File: pulsesensor/pulsesensor.c [hunk 18]
```diff
@@ -37,36 +42,39 @@ void pulsesensor_main() {
...
 
-            lastHearbeatdetected = getMillis();
+            lastHearbeatdetected = getMillis(); //Set the last time a heartbeat was detected
             bpm = 60000 / beatMsec;
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:59]: Why divide it by 60000?

File: pulsesensor/pulsesensor.c [hunk 19]
```diff
@@ -37,36 +42,39 @@ void pulsesensor_main() {
...
             beatMsec = 0;
 
-            addValue(bpm);
+            addValue(bpm);                      //Add the bpm value to the bpmValues-Array
         } else {
             if (getMillis() - lastHearbeatdetected > 7000) {
 	            clearBpm();
 	        }
         }
         beatMsec += waitTime;
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:69]: What is this execution branch for?

File: pulsesensor/pulsesensor.c (hunk 10 above)
Comment (Fred194) [reply pulsesensor/pulsesensor.c:17]: It is fine for me to use this method for the length of the array

File: pulsesensor/pulsesensor.c [hunk 20]
```diff
@@ -37,36 +42,39 @@ void pulsesensor_main() {
...
-	if(getMillis() - sendStartMillis > 1000){
+	if(getMillis() - sendStartMillis > 1000){ //To send the bpmValue every second
 		sendStartMillis = getMillis();
 		bluetooth_send_pulse(bpmValue);
 	}
 }
 
-//heartbeat-detection method inspired from the datasheet from joy-it
+//Heartbeat-detection method inspired from the datasheet from joy-it
 char heartbeatDetected(int delay, int ADCvalue) {
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:78]: Please put a description on what exactly is done how on top of this method? I don't understand what is done by purely looking at the comments

File: pulsesensor/pulsesensor.c [hunk 21]
```diff
@@ -98,40 +106,41 @@ char heartbeatDetected(int delay, int ADCvalue) {
     }
     return result;
 }
-
+//Returns the average of the bpm values
 int getBpm() {
     return average;
 }
-
+//Adds bpm value to the bpmValues-Array
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:113]: Why?

File: pulsesensor/pulsesensor.c [hunk 22]
```diff
@@ -98,40 +106,41 @@ char heartbeatDetected(int delay, int ADCvalue) {
     }
     return result;
 }
-
+//Returns the average of the bpm values
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:109]: What is this average?

File: pulsesensor/pulsesensor.c [hunk 23]
```diff
@@ -98,40 +106,41 @@ char heartbeatDetected(int delay, int ADCvalue) {
...
 int getBpm() {
     return average;
 }
-
+//Adds bpm value to the bpmValues-Array
 void addValue(int bpm) {
-    //for normalizing purpose
+    //Normalize the values - values under 40 and over 200 will not be saved in the bpmValues-Array
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:115]: Why?

File: pulsesensor/pulsesensor.c [hunk 24]
```diff
@@ -98,40 +106,41 @@ char heartbeatDetected(int delay, int ADCvalue) {
...
     if (bpm < 40 || bpm > 200) {
         return;
     }
-    //increases the position of the values by one
+    //Increases the position of the values by one
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:119]: Why?

File: pulsesensor/pulsesensor.c [hunk 25]
```diff
@@ -98,40 +106,41 @@ char heartbeatDetected(int delay, int ADCvalue) {
...
-    //calculates the average of the 10 bpm values in the array
+    //Calculates the average of the 10 bpm values in the array
     for (int i = 0; i < valueLength; i++) {
-        sum += bpmValues[i];
+        sum += bpmValues[i]; //Increases the sum with the values of the array
     }
-    if (arrayLength < valueLength) {
+    if (arrayLength < valueLength) { //Increases the arrayLength if the arrayLength is lower than the valueLength
         arrayLength++;
     }
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:133]: This is only done at the startup. Please note, that this is a special case and when and why it is needed.

File: pulsesensor/pulsesensor.c [hunk 26]
```diff
@@ -98,40 +106,41 @@ char heartbeatDetected(int delay, int ADCvalue) {
...
-    average = sum / arrayLength;
+    average = sum / arrayLength; //To calculate the average
```
Comment (Max-Leopold) [inline pulsesensor/pulsesensor.c:134]: No need for the "to"

File: pulsesensor/pulsesensor.c (hunk 18 above)
Comment (NKoepf) [reply pulsesensor/pulsesensor.c:59]: probably Conversion from millis to minutes

File: adc/adc.c [hunk 27]
```diff
@@ -6,31 +6,46 @@
 #include <avr/io.h>
 #include <avr/interrupt.h>
 
-int adc_current_value = 0;
 
+int adc_current_value = 0; //The last adc value used to find a peak to recognize a heartbeat
+
+//The ADC converts the analog input from the sensor to a digital value which we store in the variable adc_current_value
```
Comment (Max-Leopold) [inline adc/adc.c:12]: Nice

**Pull request merged**